File: LeetCode/Study Plan/Practice Problems/Easy/Linked List Cycle/Go/linked_list_cycle.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func New(val int) *ListNode { return &ListNode{Val: val, Next: nil} }

func has_cycle(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return false
	}

	sp, fp := head, head

	for fp.Next != nil && fp.Next.Next != nil {
		sp = sp.Next
		fp = fp.Next.Next
		if sp == fp {
			return true
		}
	}
	return false
}

func main() {
	l := New(3)
	l.Next = New(2)
	l.Next.Next = New(0)
	l.Next.Next.Next = New(-4)
	l.Next.Next.Next.Next = l.Next
	fmt.Println(has_cycle(l))
}

File: LeetCode/Easy/Linked List Cycle/C/linked_list_cycle.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
typedef struct ListNode ListNode;

bool hasCycle(struct ListNode *head) {
    if (head == NULL || head->next == NULL) return false;
    ListNode *sp=head, *fp=head;
    while(fp->next != NULL && fp->next->next != NULL) {
        sp = sp->next;
        fp = fp->next->next;
        if (sp == fp) return true;
    }
    return false;
}
